tests/jni_api.rs

```diff
@@ -315,122 +315,140 @@ pub fn java_byte_array_from_slice() {
-        //auto_ptr.commit().unwrap();
-    }
-
-    // Confirm modification of original Java array
-    let mut res: [i8; 3] = [0; 3];
-    env.get_byte_array_region(java_array, 0, &mut res).unwrap();
-    assert_eq!(res[0], 2);
-    assert_eq!(res[1], 3);
-    assert_eq!(res[2], 4);
-}
-
-#[test]
-pub fn get_long_array_elements() {
-    let env = attach_current_thread();
-
-    // Create original Java array
-    let buf: &[i64] = &[1, 2, 3];
-    let java_array = env
-        .new_long_array(3)
-        .expect("JNIEnv#new_long_array must create a java array with given size");
-
-    // Insert array elements
-    let _ = env.set_long_array_region(java_array, 0, buf);
-
-    // Use a scope to test Drop
-    {
-        // Get long array elements auto wrapper
-        let auto_ptr = env
-            .get_long_array_elements(java_array, ReleaseMode::CopyBack)
-            .unwrap();
-
-        // Check array size
-        assert_eq!(auto_ptr.size().unwrap(), 3);
-
-        // Check pointer access
-        let ptr = auto_ptr.as_ptr();
-        assert_eq!(unsafe { *ptr.offset(0) }, 1);
-        assert_eq!(unsafe { *ptr.offset(1) }, 2);
-        assert_eq!(unsafe { *ptr.offset(2) }, 3);
-
-        // Check pointer From access
-        let ptr: *mut jlong = std::convert::From::from(&auto_ptr);
-        assert_eq!(unsafe { *ptr.offset(0) }, 1);
-        assert_eq!(unsafe { *ptr.offset(1) }, 2);
-        assert_eq!(unsafe { *ptr.offset(2) }, 3);
-
-        // Check pointer into() access
-        let ptr: *mut jlong = (&auto_ptr).into();
-        assert_eq!(unsafe { *ptr.offset(0) }, 1);
-        assert_eq!(unsafe { *ptr.offset(1) }, 2);
-        assert_eq!(unsafe { *ptr.offset(2) }, 3);
-
-        // Modify
-        unsafe {
-            *ptr.offset(0) += 1;
-            *ptr.offset(1) += 1;
-            *ptr.offset(2) += 1;
-        }
-
-        // Commit would be necessary here, if there were no closure
-        //auto_ptr.commit().unwrap();
-    }
-
-    // Confirm modification of original Java array
-    let mut res: [i64; 3] = [0; 3];
-    env.get_long_array_region(java_array, 0, &mut res).unwrap();
-    assert_eq!(res[0], 2);
-    assert_eq!(res[1], 3);
-    assert_eq!(res[2], 4);
-}
+    };
+}
+
+test_get_array_elements!(
+    get_array_elements,
+    jint,
+    new_int_array,
+    get_int_array_region,
+    set_int_array_region
+);
+
+test_get_array_elements!(
+    get_int_array_elements,
+    jint,
+    new_int_array,
+    get_int_array_region,
+    set_int_array_region
+);
+
+test_get_array_elements!(
+    get_long_array_elements,
+    jlong,
+    new_long_array,
+    get_long_array_region,
+    set_long_array_region
+);
+
+test_get_array_elements!(
+    get_byte_array_elements,
+    jbyte,
+    new_byte_array,
+    get_byte_array_region,
+    set_byte_array_region
+);
+
+test_get_array_elements!(
+    get_boolean_array_elements,
+    jboolean,
+    new_boolean_array,
+    get_boolean_array_region,
+    set_boolean_array_region
+);
```

Comment:
Are "1", "2", "3" values for boolean well defined? What does their increment mean?

Reply:
Good point. I haven't tought about it. I'll separate that test case for clarity.

Reply:
Come to think of it, in Java boolean is well defined, as you have true and false. The same in Rust.

But in JNI land, these "booleans" are just `u8`s under the hood. So, these "1", "2", "3" values are all a redudant way to say "true".

I would leave it as it is, not to break the generality of the test macro. I can also modify the test macro to start with a zero and do one increment to 1, so that it most clearly mirrors the boolean types.

What do you think?

Reply:
I'd use a separate test for boolean, because the spec says the values are either 0 or 1: 
https://docs.oracle.com/en/java/javase/11/docs/specs/jni/types.html#primitive-types

Writing anything else is probably UB: JVM may keep the provided values; or normalize them; or fail. If the test passes, then it probably keeps them as is.

Generally, writing anything other than 0 or 1 may be problematic if some other JNI code reading these values compares them to JNI_FALSE/JNI_TRUE.

Reply:
Take a look at the latest version, and if you don't like it, I'll just write a separate test.